Extract Table of Contents from content to allow wider source view

adamwoodnz opened the pull request:
See #76 

This PR moves the Table of Contents from the content of the page and inserts it into the header.

- On wide screens it's fixed in the top right
- On medium screens it's within the header
- On small screens it's at the top of the page

A further improvement might be to add expand/collapse behaviour and collapse by default on small screens. Although that might mask the utility of it for jumping down a long page.

Wide
<img width="1700" alt="Screen Shot 2022-06-08 at 2 34 38 PM" src="https://user-images.githubusercontent.com/1017872/172519005-9e4ab36e-5ac1-429e-9e9a-fd63e018c5a4.png">

Medium
<img width="1025" alt="Screen Shot 2022-06-10 at 11 10 45 AM" src="https://user-images.githubusercontent.com/1017872/172959977-ffe5a6e9-568d-4bb5-8f36-f934290fac3a.png">

Small
<img width="480" alt="Screen Shot 2022-06-10 at 11 09 18 AM" src="https://user-images.githubusercontent.com/1017872/172959865-86819289-c7cf-410e-be82-d80b4a7985c4.png">

adamwoodnz committed:
Extract Table of Contents from content and fix it to the right
- On medium screens position it within the header
- On small screens hide it from view

dd32 commented on source/wp-content/plugins/handbook/inc/table-of-contents.php:
```diff
@@ -164,7 +164,7 @@ public function add_toc( $content ) {
 		$toc .= "</ul>\n</div>\n";
 
-		return $toc . $content;
+		return array('toc' => $toc, 'content' => $content);
```
My suggestion would be to rename this from `add_toc` to something like `parse_content`, then add a new `add_toc` method which is similar to:
```
function add_toc( $content ) {
   $parts = $this->parse_content( $content );
   return $parts['toc'] . $parts['content'];
}
```

That way the function syntax remains unchanged for other uses of the plugin.. and for this filter in the plugin: `add_filter( 'the_content', array( $this, 'add_toc' ) );` which would expect a stringy return.

adamwoodnz replied:
Done! Thanks for the suggestion. Tested that this works with the old `content-reference.php`

adamwoodnz committed:
Refactor table of contents for backwards compatibility

adamwoodnz commented on source/wp-content/plugins/handbook/inc/table-of-contents.php:
```diff
@@ -138,12 +151,11 @@ public function add_toc( $content ) {
+			return array('content' => $content, 'toc' => null);
 		}
 
-		$contents_header = 'h' . reset( $items )['level']; // Duplicate the first <h#> tag in the document for the TOC header
```
The header shouldn't be an `h` tag now that the ToC is before the H1

adamwoodnz committed:
Move ToC to start of content for accessibility
- Display top of page on small screens
- Inline to right in header on medium screens

adamwoodnz committed:
Revert doc change

StevenDufresne commented:
I wonder if we should keep it closer to the content as opposed to the browser's edge:

```
// Something like
right: calc(((100vw - 60em ) / 2 ) - ( 250px - 15px )  )
```
[See Example](https://d.pr/v/KhFSme.mp4)

adamwoodnz commented:
> I wonder if we should keep it closer to the content as opposed to the browser's edge:
> 
> ```
> // Something like
> right: calc(((100vw - 60em ) / 2 ) - ( 250px - 15px )  )
> ```
> 
> [See Example](https://d.pr/v/KhFSme.mp4)

Yeah I think that's a good call

tellyworth commented:
Sounds like this is broken on other handbook pages currently. Is it easily fixable or should we punt this from the milestone?

iandunn commented:
> On medium screens it's within the header

Other handbooks have variable-length ToCs, and some are much longer:

* https://developer.wordpress.org/plugins/wordpress-org/detailed-plugin-guidelines/
* https://developer.wordpress.org/rest-api/frequently-asked-questions/
* https://developer.wordpress.org/themes/basics/template-hierarchy/
* https://developer.wordpress.org/coding-standards/wordpress-coding-standards/php/

![Screen Shot 2022-06-10 at 9 12 36 AM](https://user-images.githubusercontent.com/484068/173107787-eb3768d0-6433-416a-b33b-e4424e24fdb4.png)

On those, it might look better to make it take 100% width of `main.site-main`, like it does for mobile.

adamwoodnz commented:
> Sounds like this is broken on other handbook pages currently. Is it easily fixable or should we punt this from the milestone?

Correct, the other pages are still calling `add_toc` directly via this [filter](https://github.com/WordPress/wporg-developer/blob/trunk/source/wp-content/plugins/handbook/inc/table-of-contents.php#L85):
```
        public function load_filters() {
		if ( is_singular( $this->post_types ) && ! is_embed() ) {
			add_filter( 'the_content', array( $this, 'add_toc' ) );
		}
	}
```

The result is that on medium screens where the ToC is displayed inline the layout is affected.

<img width="1220" alt="Screen Shot 2022-06-15 at 2 18 08 PM" src="https://user-images.githubusercontent.com/1017872/173722470-90715a6f-7cac-4213-af40-b3e12b8eef48.png">

For accessibility reasons my change in the reference handbook involves moving the ToC before the H1 and adding some layout markup.

`content_handbook.php` simply calls these functions:
```
<h1><?php the_title(); ?></h1>

<?php
/*
 * Content
 */
?>

<?php the_content( __( '(More ...)' , 'wporg' ) ); ?>
```

So I've been trying to do something using filters for `the_content` and `the_title` to split the ToC parts and render them separately. The content filter works, but I've hit a wall trying to get the ToC part within `the_title`:

```
        public function load_filters() {
		if ( is_singular( $this->post_types ) && ! is_embed() ) {
			add_filter( 'the_content', array( $this, 'content_with_links' ) );
			add_filter( 'the_title', array( $this, 'title_with_toc' ) );
		}
	}

	function content_with_links( $content ) {
		$parts = $this->parse_content( $content );

		return $parts['content'];
	}

	function title_with_toc( $title ) {
		// How do I get the content?
		$parts = $this->parse_content( $content );

		return $title . $parts['toc'];
	}
```

I did try saving the $toc value on the class instance but it still wan't available in `the_title`, I guess due to lifecycle but my WordPress knowledge is limited here.

EDIT:
Potentially moving the ToC above the H1 (away from the content) is not necessary. @dd32 you originally pointed to this doc https://www.w3.org/TR/WCAG20-TECHS/G64.html for guidance but I don't see a requirement there. Did you read this somewhere else perhaps?

adamwoodnz commented:
> > On medium screens it's within the header
> 
> Other handbooks have variable-length ToCs, and some are much longer:
> 
> * https://developer.wordpress.org/plugins/wordpress-org/detailed-plugin-guidelines/
> * https://developer.wordpress.org/rest-api/frequently-asked-questions/
> * https://developer.wordpress.org/themes/basics/template-hierarchy/
> * https://developer.wordpress.org/coding-standards/wordpress-coding-standards/php/
> 
> ![Screen Shot 2022-06-10 at 9 12 36 AM](https://user-images.githubusercontent.com/484068/173107787-eb3768d0-6433-416a-b33b-e4424e24fdb4.png)
> 
> On those, it might look better to make it take 100% width of `main.site-main`, like it does for mobile.

👍 This is broken on this page but I have made it full width on the Content Reference page

adamwoodnz commented:
Closed in favor of https://github.com/WordPress/wporg-developer/pull/122
